app/src/main/java/com/github/lookupgroup27/lookup/ui/feed/Feed.kt

```diff
@@ -87,101 +96,142 @@ fun FeedScreen(
+                              .testTag(stringResource(R.string.loading_indicator_test_tag)),
+                      contentAlignment = Alignment.Center) {
+                        if (!locationPermissionGranted) {
+                          Text(
+                              text = stringResource(R.string.location_permission_required),
+                              style = MaterialTheme.typography.bodyLarge.copy(color = Color.White))
+                        } else {
+                          CircularProgressIndicator(color = Color.White)
+                        }
+                      }
+                } else {
+                  LazyColumn(
+                      modifier = Modifier.fillMaxSize(),
+                      contentPadding = PaddingValues(vertical = 16.dp),
+                      verticalArrangement = Arrangement.spacedBy(30.dp)) {
+                        items(nearbyPosts) { post ->
+                          PostItem(
+                              post = post,
+                              starStates =
+                                  postRatings[post.uid] ?: mutableListOf(false, false, false),
+                              onRatingChanged = { newRating ->
+                                val oldPostRatings =
+                                    postRatings[post.uid] ?: mutableListOf(false, false, false)
+                                val oldStarCounts = oldPostRatings.count { it }
+                                postRatings[post.uid] = newRating.toList()
+                                val starsCount = newRating.count { it }
 
-                        postsViewModel.updatePost(
-                            post.copy(
-                                averageStars = newAvg,
-                                starsCount = newStarsCount,
-                                usersNumber = newUsersNumber,
-                                ratedBy =
-                                    if (!isReturningUser) {
-                                      post.ratedBy + userEmail
+                                val updatedRatings = profile?.ratings?.toMutableMap()
+                                updatedRatings?.set(post.uid, starsCount)
+                                val newProfile: UserProfile =
+                                    profile?.copy(
+                                        username = username,
+                                        bio = bio,
+                                        email = email,
+                                        ratings = updatedRatings ?: emptyMap())
+                                        ?: UserProfile(
+                                            username = username,
+                                            bio = bio,
+                                            email = email,
+                                            ratings = updatedRatings ?: emptyMap())
+                                profileViewModel.updateUserProfile(newProfile)
+
```

Comment:
UserProfile construction and copying logic are repeated. It can be extracted into a helper function for reuse:
Suggested code :
```
fun updateProfileRatings(
    profile: UserProfile?,
    postUid: String,
    starsCount: Int
): UserProfile {
    val updatedRatings = profile?.ratings?.toMutableMap()?.apply {
        this[postUid] = starsCount
    } ?: mapOf(postUid to starsCount)
    return profile?.copy(ratings = updatedRatings) ?: UserProfile(
        username = username,
        bio = bio,
        email = email,
        ratings = updatedRatings
    )
}

// Usage:
val newProfile = updateProfileRatings(profile, post.uid, starsCount)
profileViewModel.updateUserProfile(newProfile)

```